# SDA Assignment 2

Ross Herencia  


In [1]:
import numpy as np

### 2.1  
The three datasets are

In [2]:
x = [-0.3, 0.2, -0.1, 0.3, 0.2]
y = [0.7, 0.6, 0.6, 0.2, 0.0]
z = [-0.8, -0.5, 0.6, 0.3, 0.3]

The mean of each dataset may be calculated using 
$$ \bar{x} = \frac{1}{N} \sum_{i=1}^{N} x_i $$

In [3]:
def mean(data):
    total = 0
    for i in data:
        total += i
    mean = total / len(data)
    return mean

In [4]:
mean_x = mean(x)
mean_y = mean(y)
mean_z = mean(z)
print('mean of x:', mean_x)
print('Mean of y:', mean_y)
print('Mean of z:', mean_z)

mean of x: 0.06000000000000001
Mean of y: 0.42000000000000004
Mean of z: -0.020000000000000018


$$ \therefore~\bar{x} = 0.06,\\
\bar{y} = 0.042,\\
\bar{z} = -0.02,\\ $$
to 2 d.p. I will keep the outputs to more decimal places so that they can be carried forward in the following calculations.

The Bessel corrected standard deviation is given by  

$$ \sigma_x = \sqrt{\frac{1}{n-1} \sum_{i=1}^{N} (x_i - \bar{x})^2} $$

In [5]:
def std(data, mean):
    total = 0
    for i in data:
        total += (i - mean)**2
    std = (total / (len(data) - 1))**0.5
    return std

In [6]:
std_x = std(x, mean_x)
std_y = std(y, mean_y)
std_z = std(z, mean_z)
print('Standard deviation of x:', std_x)
print('Standard deviation of y:', std_y)
print('Standard deviation of z:', std_z)

Standard deviation of x: 0.25099800796022265
Standard deviation of y: 0.30331501776206204
Standard deviation of z: 0.5974947698515862


$$ \therefore~ \sigma_x = 0.25,\\
\sigma_y = 0.30,\\
\sigma_z = 0.60. $$

The elements of the covariance matrix are given by  

$$ V = \sum_{i,j} \sigma_{ij}. $$  

The Bessel corrected covariances are calculated using

$$ \sigma_{xy} = \frac{1}{N-1} \sum_{i=1}^{N} (x_i - \bar{x})(y_i - \bar{y}), $$  

and clearly $\sigma_{xy} = \sigma_{yx}$.

In [7]:
def covariance(data_1, data_2, mean_1, mean_2):
    total = 0
    for i in range(len(data_1)):
        total += (data_1[i] - mean_1)*(data_2[i] - mean_2)
    cov = total / (len(data_1) - 1)
    return cov

In [8]:
cov_xy = covariance(x, y, mean_x, mean_y)
cov_xz = covariance(x, z, mean_x, mean_z)
cov_yz = covariance(y, z, mean_y, mean_z)
print('Covariance of x,y:', cov_xy)
print('Covariance of x,z:', cov_xz)
print('Covariance of y,z:', cov_yz)

Covariance of x,y: -0.054
Covariance of x,z: 0.05899999999999999
Covariance of y,z: -0.09949999999999999


Therefore,  

$$\sigma_{x}^2 = 0.063,~\sigma_{y}^2 = 0.09,~\sigma_{z}^2 = 0.36,\\  
\sigma_{xy} = -0.05,~\sigma_{xz} = 0.06,~\sigma_{yz} = -0.10 $$

$$ \implies V = \begin{pmatrix}
\sigma_x^2 & \sigma_{xy} & \sigma_{xz}\\
\sigma_{yx} & \sigma_y^2 & \sigma_{yz}\\
\sigma_{zx} & \sigma_{zy} &\sigma_z^2 \end{pmatrix} = \begin{pmatrix}
0.063 & -0.05 & 0.06\\
-0.05 & 0.09 & -0.10\\
0.06 & -0.10 & 0.36
\end{pmatrix}_. $$

The correlation matrix elements are 

$$ \rho_{ij} = \frac{\sigma_{ij}}{\sigma_i \sigma_j} $$
so if $i = j, \rho_{ii} = 1$.

In [9]:
rho_xy = cov_xy / (std_x * std_y)
rho_xz = cov_xz / (std_x * std_z)
rho_yz = cov_yz / (std_y * std_z)
print('rho_xy = ', rho_xy)
print('rho_xz = ', rho_xz)
print('rho_yz = ', rho_yz)

rho_xy =  -0.7092993656151906
rho_xz =  0.3934120236078694
rho_yz =  -0.5490287146712387


$$ \therefore ~ \rho_{xy} = -0.71,\\
\rho_{xz} = 0.39,\\
\rho_{yz} = -0.55. $$

So the correlation martix is 

$$ \rho = \begin{pmatrix}
1 & -0.71 & 0.39\\
-0.71 & 1 & -0.55\\
0.39 & -0.55 & 1
\end{pmatrix} $$

Spearman's ranked correlation is 

$$ \rho = 1 - \frac{6}{N(N^2 - 1)}\sum_{i=1}^N d_i^2, $$

where $d_i^2$ is the difference in rank between $x_i$ and $y_i$.

In [10]:
x_rank = [(1, -0.3), (3.5, 0.2), (2, -0.1), (5, 0.3), (3.5, 0.2)] # (rank, value)
y_rank = [(5, 0.7), (3.5, 0.6), (3.5, 0.6), (2, 0.2), (1, 0.0)]

Where there is a tie in the rank (e.g. y = 0.6 occurs twice), the average of current rank (r) and the next (r + 1)... is taken for all tied values.

In [11]:
def spearman(x_rank, y_rank):
    total = 0
    for i,j in zip(x_rank, y_rank):
        total += (i[0] - j[0])**2
    s_correlation = 1 - (6 * total) / (len(x_rank) * (len(x_rank)**2 - 1))
    return s_correlation

In [12]:
spearman(x_rank, y_rank)

-0.675

So, the Spearman ranked correlation for x and y is $ \rho = -0.68 . $

### 2.2  

Given the PDF mean $\mu$  
$$ \widehat{V{(x)}} = \frac{1}{N} \sum_i^N (x_i - \mu)^2,$$
where $N = |\Omega|$.  
In the limit that $ N \rightarrow \infty$ then $\widehat{V{(x)}} \rightarrow V(x)$ because $\mu$ is the mean of the PDF therefore, the estimate is consistent.  
The Bessel correction is no longer required beacuse the estimator is not biased.

### 2.3  
$$ A_1 = 1.2 \pm 0.3,\\
B_1 = -0.8 \pm 0.2,\\
A_2 = 0.9 \pm 0.1,\\
B_2 = -0.5 \pm 0.2,\\
\rho_{AB}^{(1)} = 0.2,\\
\rho_{AB}^{(2)} = 0.1.
$$
The weighted average of A and B is given by the formulae

$$ \bar{x} = \left[\sum_{j=1}^{2} \begin{pmatrix}
\sigma_A^2 & \rho_{AB} \sigma_{A} \sigma_{B}\\
\rho_{AB} \sigma_{A} \sigma_{B} & \sigma_B^2 \end{pmatrix}_j^{-1}\right]  \bullet
\left[\sum_{j=1}^{2} \begin{pmatrix} 
\sigma_A^2 & \rho_{AB} \sigma_{A} \sigma_{B}\\
\rho_{AB} \sigma_{A} \sigma_{B} & \sigma_B^2 \end{pmatrix}_j^{-1} 
\begin{pmatrix} 
A\\
B \end{pmatrix}_j \right] $$  

and 

$$ V = \left[\sum_{j=1}^{2} \begin{pmatrix}
\sigma_A^2 & \rho_{AB} \sigma_{A} \sigma_{B}\\
\rho_{AB} \sigma_{A} \sigma_{B} & \sigma_B^2 \end{pmatrix}_j^{-1}\right], $$

since $\rho_{AB} = \sigma_{AB}~/~\sigma_A \sigma_B$.

In [13]:
m_1 = np.matrix('0.09, 0.012; 0.012, 0.04') # Measurement 1 matrix
print('Measurement 1 matrix of covariances:\n', m_1)

m_2 = np.matrix('0.01, 0.002; 0.002, 0.04')
print('Measurement 2 matrix of covariances:\n', m_2)

Measurement 1 matrix of covariances:
 [[0.09  0.012]
 [0.012 0.04 ]]
Measurement 2 matrix of covariances:
 [[0.01  0.002]
 [0.002 0.04 ]]


In [14]:
v_1 = np.matrix([[1.2], [-0.8]]) # vector of measurements
print('Measurement 1 values:\n', v_1)
v_2 = np.matrix([[0.9], [-0.5]])
print('Measurement 2 values:\n', v_2)

Measurement 1 values:
 [[ 1.2]
 [-0.8]]
Measurement 2 values:
 [[ 0.9]
 [-0.5]]


We need to invert these matrices which can be done using 

$$ \begin{pmatrix}
\sigma_A^2 & \rho_{AB} \sigma_{A} \sigma_{B}\\
\rho_{AB} \sigma_{A} \sigma_{B} & \sigma_B^2 \end{pmatrix}^{-1} = \frac{1}{\sigma_A^2 \sigma_B^2 - (\rho_{AB} \sigma_{A} \sigma_{B})^2} \begin{pmatrix}
\sigma_B^2 & -\sigma_{AB}\\
-\sigma_{AB} & \sigma_A^2 \end{pmatrix}$$

In [15]:
def inverse(m):
    coeff = (m[0,0] * m[1,1] - m[1,0] * m[0,1])**(-1)
    inverted = np.matrix([[m[1,1], -m[1,0]], [-m[0,1], m[0,0]]])
    return coeff*inverted

In [16]:
m_1_inv = inverse(m_1)
print('m_1 inverted:\n', m_1_inv)
m_2_inv = inverse(m_2)
print('m_2 inverted:\n', m_2_inv)

m_1 inverted:
 [[11.57407407 -3.47222222]
 [-3.47222222 26.04166667]]
m_2 inverted:
 [[101.01010101  -5.05050505]
 [ -5.05050505  25.25252525]]


In [17]:
sum_12_inv = m_1_inv + m_2_inv
sum_12_inv

matrix([[112.58417508,  -8.52272727],
        [ -8.52272727,  51.29419192]])

In [18]:
term_1 = inverse(sum_12_inv)
term_1  # This is V!

matrix([[0.00899539, 0.00149462],
        [0.00149462, 0.01974372]])

In [19]:
term_2 = m_1_inv * v_1 + m_2_inv * v_2
term_2

matrix([[110.1010101 ],
        [-42.17171717]])

In [20]:
x_bar = term_1 * term_2
x_bar

matrix([[ 0.92737058],
        [-0.66806766]])

$$ \therefore V = \begin{pmatrix}
0.0090 & 0.0015\\
0.0015 & 0.020 \end{pmatrix} $$  
and  
$$ \bar{x} = \begin{pmatrix}
0.927\\
-0.668 \end{pmatrix}$$

$$ \implies \bar{A} = 0.93 \pm 0.09,\\
\bar{B} = -0.67 \pm 0.14 $$
and $$\sigma_{AB} = 0.0015.$$

Now, setting $\rho_{AB} = 0$ for measurements 1 and 2 we get

In [21]:
m_10 = np.matrix('0.09, 0; 0, 0.04') # Measurement 1 matrix
print('Measurement 1 matrix of covariances:\n', m_10)

m_20 = np.matrix('0.01, 0; 0, 0.04')
print('Measurement 2 matrix of covariances:\n', m_20)

Measurement 1 matrix of covariances:
 [[0.09 0.  ]
 [0.   0.04]]
Measurement 2 matrix of covariances:
 [[0.01 0.  ]
 [0.   0.04]]


In [22]:
m_10_inv = inverse(m_10)
print('m_10 inverted:\n', m_10_inv)
m_20_inv = inverse(m_20)
print('m_20 inverted:\n', m_20_inv)

m_10 inverted:
 [[11.11111111 -0.        ]
 [-0.         25.        ]]
m_20 inverted:
 [[100.  -0.]
 [ -0.  25.]]


In [23]:
sum_120_inv = m_10_inv + m_20_inv

term_10 = inverse(sum_120_inv)
print('term_10:\n', term_10)

term_20 = m_10_inv * v_1 + m_20_inv * v_2
print('term_20:\n', term_20)

x_bar0 = term_10 * term_20
print('x_bar:\n', x_bar0)

term_10:
 [[0.009 0.   ]
 [0.    0.02 ]]
term_20:
 [[103.33333333]
 [-32.5       ]]
x_bar:
 [[ 0.93]
 [-0.65]]


$$ \therefore V = \begin{pmatrix}
0.009 & 0\\
0 & 0.020 \end{pmatrix} $$  
and  
$$ \bar{x} = \begin{pmatrix}
0.93\\
-0.65 \end{pmatrix}$$

$$ \implies \bar{A} = 0.93 \pm 0.09,\\
\bar{B} = -0.65 \pm 0.14$$
and $$\sigma_{AB} = 0.$$

Separate averages for A and B are given by

$$ \bar{x} = \frac{x_1/\sigma_1^2 + x_2/\sigma_2^2}{1/\sigma_1^2 + 1/\sigma_2^2}$$
and 
$$ \sigma_x = (1/\sigma_1^2 + 1/\sigma_2^2)^{-1/2}$$

In [24]:
def average(x1, x2, var1, var2):
    numerator = ((x1 / var1) + (x2 / var2))
    uncertainty = ( var1**-1 + var2**-1)
    avg = numerator / uncertainty
    output = (avg, uncertainty**-0.5)
    return output

In [25]:
average(v_1[0,0], v_2[0,0], m_1[0,0], m_2[0,0])

(0.9299999999999999, 0.09486832980505137)

In [26]:
average(v_1[1,0], v_2[1,0], m_1[1,1], m_2[1,1])

(-0.65, 0.1414213562373095)

$$ \implies \bar{A} = 0.93 \pm 0.09\\
\bar{B} = -0.65 \pm 0.14$$

Therefore, these results agree with setting $\rho_{AB} = 0$. This makes sense  because that constraint implies that A and B are not correlated, so they are independent, which is exactly the case that you would use separate averages.